pkg/controller/operators/catalog/step_ensurer.go

```diff
@@ -116,12 +116,35 @@ func (o *StepEnsurer) EnsureServiceAccount(namespace string, sa *corev1.ServiceA
 		return
 	}
 
+	// Before UpdateServiceAccount and dereference secrets of the older SA,
```

Comment:
Doesn't the original bug report have this issue with the same serviceaccount name? 

From line 114, this section should only be hit if the serviceaccount doesn't exist at all, so I'm not convinced this will fix the bug.

I think the issue may be that `Create` on `ServiceAccount` always creates a secret even if the serviceaccount exists? Maybe we should start with a `Get` before the initial `Create`?

Do you agree, or am I missing how this fixes the issue?

Reply:
You might have missed `!k8serrors.IsAlreadyExists` which reports any SA creation problem other than SA exists. Therefore, the only scenario where *my addition* hits is if the SA **does** exists. 

The logic of the original code got me at first as well.

Reply:
This is what I get for reviewing too quickly 😄  Ignore my previous comment.

However - I really don't think we should be in the business of deleting the serviceaccount secrets.

I did some testing, and the issue is that when we `Update`, our SA definition doesn't have the secrets set, which removes the configured secret and causes the SA controller to generate a new one:

```sh
$ k -n default create serviceaccount test
serviceaccount/test created

$ k -n default get secrets
NAME                  TYPE                                  DATA   AGE
default-token-67bw8   kubernetes.io/service-account-token   3      3m23s
test-token-9lx4p      kubernetes.io/service-account-token   3      3s

# an update that doesn't cause the bug
$ k -n default label serviceaccounts test with=test
serviceaccount/test labeled

# no duplicate secrets
$ k -n default get secrets
NAME                  TYPE                                  DATA   AGE
default-token-67bw8   kubernetes.io/service-account-token   3      6m40s
test-token-9lx4p      kubernetes.io/service-account-token   3      3m20s

# an update that causes the bug - remove the secrets block
$ k -n default edit serviceaccounts test
serviceaccount/test edited

# now there are duplicates
$ k -n default get secrets
NAME                  TYPE                                  DATA   AGE
default-token-67bw8   kubernetes.io/service-account-token   3      8m3s
test-token-9lx4p      kubernetes.io/service-account-token   3      4m43s
test-token-xwmpg      kubernetes.io/service-account-token   3      4s
```

The solution here shouldn't be to delete the extra secrets, it should be to not create the extra secrets in the first place.

A quick option for now would be to:

- Get the current ServiceAccount (if it exists)
- Add the current secret list to the object from the step
- Then call update with that object - it will have the secret list so a new one won't be generated

But it would be even better to use `DeepDerivative` or server-side apply.

Reply:
That was my other thought on the fix too. I eventually chose this to make sure we always have the right token in the secrets. I worry that secrets may change in the install plan rendering old secrets to be invalid. However I might be worrying for nothing. I’ll make a commit to use that approach.

Reply:
@ecordell I liked the server-side approach but it being in beta gives me goosebumps. I am not entirely sure about your suggestion about `DeepDerivative`. If you are talking about https://github.com/kubernetes/kubernetes/blob/6be19d85f206d64044ae940540a437c65675a526/third_party/forked/golang/reflect/deep_equal.go#L378  then I do not see how we can use it. If you are talking about the concept. Not every field can be copied (ie: timestamp, etc). If we copy every unset field in the new SA from the existing SA, we need to know exactly what is not copiable. I do not think my understanding of your `DeepDerivative` is correct. Therefore, please let me know what you are explicitly suggesting. Thank you!